### PR TITLE
added funlen on new commits

### PR DESCRIPTION
**Pull request by facchettos**
### Which issue this PR addresses:
This PR enables the `funlen` linter on new commits (beginning from the parent commit of that PR).
This linter will complain when we have functions that are too long, but can be disabled with `//nolint:funlen` if this is a deliberate choice. 
Having shorter functions makes it easy to have better testing coverage.  

Fixes

### What this PR does / why we need it:
This PR enables the `funlen` linter on new commits (beginning from the parent commit of that PR).
This linter will complain when we have functions that are too long, but can be disabled with `//nolint:funlen` if this is a deliberate choice. 
Having shorter functions makes it easy to have better testing coverage.  

### Test plan for issue:
No test needed, if CI passes, it should be good to go

### CODE REVIEW

File: .golangci-funlen.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+issues:
+  exclude-rules:
+  - path: _test\.go
+    linters:
+      - funlen
+  - path: pkg/swagger/\*\.go
+    linters:
+      - funlen
+  new-from-rev: 5668a316afae97b690871edac2c59eb8f84efacb
```
Comment (s-amann) [inline .golangci-funlen.yml:15]: Will there be a separate PR enforcing the linter on the current files?

File: .golangci-funlen.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  exclude-use-default: false
```
[the remark this answers is not in the document]
Comment (m1kola) [reply .golangci-funlen.yml:16]: Also - I would like to see few examples of functions. Could you please split some existing code which currently fails `funlen` without `new-from-rev`?

I would like to see how readability of the following functions improve (or not).

```
pkg/cluster/deploystorage_resources.go:265: Function 'networkInternalLoadBalancer' is too long (170 > 60) (funlen)
func (m *manager) networkInternalLoadBalancer(installConfig *installconfig.InstallConfig) *arm.Resource {
```

I suspect a lot of functions under `pkg/cluster` which return `*arm.Resource` will have to be marked with `//nolint:funlen`.


```
pkg/deploy/generator/templates_rp.go:21: Function 'rpTemplate' is too long (129 > 60) (funlen)
func (g *generator) rpTemplate() *arm.Template {

pkg/portal/ssh/proxy.go:82: Function 'newConn' has too many statements (48 > 40) (funlen)
func (s *ssh) newConn(ctx context.Context, clientConn net.Conn) error {

pkg/monitor/worker.go:47: Function 'changefeed' has too many statements (42 > 40) (funlen)
func (mon *monitor) changefeed(ctx context.Context, baseLog *logrus.Entry, stop <-chan struct{}) {

pkg/cluster/delete.go:148: Function 'deleteResources' has too many statements (48 > 40) (funlen)
func (m *manager) deleteResources(ctx context.Context) error {
```

It might be a good idea to create a separate PR for these functions. We can merge refactoring of these functions independently from this PR. We can merge refactoring even if we decide not to proceed with enabling funlen.

**Review by m1kola — COMMENTED** [review]
I left few questions and suggestions.

I personally do not like this kind of linting. I find it more irritating than helpful: when I work with codebase with linting on length I constantly find myself either needlessly splitting perfectly readable code which is slifhtly beyond limit or adding exclusions (likes of `//nolint:funlen`). I've never been in a team where we lasted with this kind of linting for more than 3 months. But I'm happy to try again :)

File: Makefile [hunk 3]
```diff
@@ -166,6 +166,7 @@ unit-test-go:
 
 lint-go:
 	go run ./vendor/github.com/golangci/golangci-lint/cmd/golangci-lint run
+	go run ./vendor/github.com/golangci/golangci-lint/cmd/golangci-lint run -c ./.golangci-funlen.yml
```
[the remark this answers is not in the document]
Comment (facchettos) [reply Makefile:169]: yes, I could not find the right combo to run some with the `new-from-rev` and some without

File: .golangci-funlen.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (facchettos) [reply .golangci-funlen.yml:16]: > Also - I would like to see few examples of functions. Could you please split some existing code which currently fails funlen without new-from-rev?
I would like to see how readability of the following functions improve (or not).

We can also tweak the settings and keep only the number of statements rather than the number of lines and the number of statements. For init functions with big structs , the number of lines is irrelevant, but the number of statement is relevant in most cases.  

Actualyl , now that I write this, I become convinced that we should disable the check for number of lines and only keep the number of statements.

[the remark this answers is not in the document]
Comment (facchettos) [reply .golangci-funlen.yml]: Also, having smaller functions is not so much about having readable code, even though it help, it's about testing. When you have bigger functions, it's much harder to test. And I'm not talking about code coverage here, because you can have code coverage with absolutely 0 relevance. Having relevant coverage is much, much harder when you have long functions than when you have smaller ones.

[the remark this answers is not in the document]
Comment (facchettos) [reply .golangci-funlen.yml:16]: > Because of new-from-rev I suspect that we will have issues with other generated code and issues only manifest themselves on the next change to the code. For example we do not change pkg/client/services/redhatopenshift/mgmt/ very often. This will probably be annoying to a person working on new clients.

This is just a matter of finding the right exclusion rules, isn't it ? If you end up with generated code that does not meet the standard, exclude it and end of the story. Or am I missing something ?

**Review by s-amann — APPROVED** [review]
